### PR TITLE
Ignore find_with_equal_func for ListStore until a suitable API is provided by the C library

### LINKED ISSUE #202: `find_with_equal_func` is not implemented for ListStore
The bindings were not generated for this function.

### PR DESCRIPTION
**Pull request by keltrycroft**
Fixes #202.

### CODE REVIEW

File: gio/src/list_store.rs [hunk 1]
```diff
@@ -4,6 +4,9 @@ use crate::ListStore;
 use glib::translate::*;
 use glib::{IsA, Object};
 use std::cmp::Ordering;
+#[cfg(any(feature = "v2_64", feature = "dox"))]
+#[cfg_attr(feature = "dox", doc(cfg(feature = "v2_64")))]
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:8]: Please remove this line.

File: gio/src/list_store.rs [hunk 2]
```diff
@@ -15,6 +18,15 @@ pub trait ListStoreExtManual {
 
     #[doc(alias = "g_list_store_sort")]
     fn sort<F: FnMut(&Object, &Object) -> Ordering>(&self, compare_func: F);
+
+    #[cfg(any(feature = "v2_64", feature = "dox"))]
+    #[cfg_attr(feature = "dox", doc(cfg(feature = "v2_64")))]
+    #[doc(alias = "g_list_store_find_with_equal_func")]
+    fn find_with_equal_func<P: IsA<glib::Object>, F: Fn(&Object, &Object) -> glib::ffi::gboolean>(
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:25]: Shouldn't this be:

```suggestion
    fn find_with_equal_func<P: IsA<glib::Object>, F: Fn(&IsA<Object>, &IsA<Object>) -> glib::ffi::gboolean>(
```


?

File: gio/src/list_store.rs [hunk 3]
```diff
@@ -52,6 +64,28 @@ impl<O: IsA<ListStore>> ListStoreExtManual for O {
...
+            let mut position = mem::MaybeUninit::uninit();
+            let ret = from_glib(ffi::g_list_store_find_with_equal_func(
+                self.as_ref().to_glib_none().0,
+                item.as_ref().to_glib_none().0,
+                Some(equal_func_trampoline),
+                position.as_mut_ptr(),
+            ));
+            let position = position.assume_init();
+            if ret {
+                Some(position)
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:83]:
```suggestion
            if ret {
                let position = position.assume_init();
                Some(position)
```

File: gio/src/list_store.rs [hunk 4]
```diff
@@ -70,3 +104,17 @@ unsafe extern "C" fn compare_func_trampoline(
         Ordering::Greater => 1,
     }
 }
+
+unsafe extern "C" fn equal_func_trampoline(
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:108]: @sdroege Should we move this function into `find_with_equal_func` as well? I thought all callbacks were directly declared inside the block calling them but maybe I'm wrong there...

File: gio/src/list_store.rs (hunk 3 above)
Comment (bilelmoussaoui) [reply gio/src/list_store.rs:83]: Just `Some(position.assume_init())`  should be fine :-)

File: gio/src/list_store.rs (hunk 4 above)
Comment (bilelmoussaoui) [reply gio/src/list_store.rs:108]: They should unless they are re-used elsewhere

File: gio/src/list_store.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply gio/src/list_store.rs:25]: It also should be returning a normal bool and not a glib:ffi::gboolean

File: gio/src/list_store.rs [hunk 5]
```diff
@@ -52,6 +64,28 @@ impl<O: IsA<ListStore>> ListStoreExtManual for O {
             )
         }
     }
+
+    fn find_with_equal_func<P: IsA<glib::Object>, F: Fn(&Object, &Object) -> glib::ffi::gboolean>(
```
Comment (bilelmoussaoui) [inline gio/src/list_store.rs:68]:
```suggestion
    #[cfg(any(feature = "v2_64", feature = "dox"))]
    #[cfg_attr(feature = "dox", doc(cfg(feature = "v2_64")))]
    fn find_with_equal_func<P: IsA<glib::Object>, F: Fn(&Object, &Object) -> glib::ffi::gboolean>(
```

File: gio/src/list_store.rs [hunk 6]
```diff
@@ -52,6 +64,28 @@ impl<O: IsA<ListStore>> ListStoreExtManual for O {
...
+        &self,
+        item: &P,
+        equal_func: F,
+    ) -> Option<u32> {
+        unsafe {
+            let mut position = mem::MaybeUninit::uninit();
```
Comment (bilelmoussaoui) [inline gio/src/list_store.rs:74]:
```suggestion
            let mut position = std::mem::MaybeUninit::uninit();
```
and drop the mem import you have at the top of the file

File: gio/src/list_store.rs (hunk 4 above)
Comment (keltrycroft) [reply gio/src/list_store.rs]: Moved callback inside block as it is not re-used elsewhere

File: gio/src/list_store.rs [hunk 7]
```diff
@@ -5,6 +5,7 @@ use glib::translate::*;
 use glib::{IsA, Object};
 use std::cmp::Ordering;
 
+
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:8]: Why this new line?

File: gio/src/list_store.rs [hunk 8]
```diff
@@ -52,6 +61,41 @@ impl<O: IsA<ListStore>> ListStoreExtManual for O {
...
+            b: glib::ffi::gconstpointer,
+        ) -> glib::ffi::gboolean {
+            let a: Borrowed<glib::Object> = from_glib_borrow(a as *mut glib::gobject_ffi::GObject);
+            let b: Borrowed<glib::Object> = from_glib_borrow(b as *mut glib::gobject_ffi::GObject);
+        
+            if *a == *b {
+                true.into_glib()
+            } else {
+                false.into_glib()
+            }
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:82]:
```suggestion
            (*a == *b).into_glib()
```

File: gio/src/list_store.rs [hunk 9]
```diff
@@ -70,3 +114,5 @@ unsafe extern "C" fn compare_func_trampoline(
         Ordering::Greater => 1,
     }
 }
+
```
Comment (GuillaumeGomez) [inline gio/src/list_store.rs:117]: Why adding those lines?

File: gio/src/list_store.rs (hunk 7 above)
Comment (keltrycroft) [reply gio/src/list_store.rs]: My mistake, no reason

File: gio/src/list_store.rs (hunk 8 above)
Comment (keltrycroft) [reply gio/src/list_store.rs]: Much cleaner, thanks

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks good to me, thanks! Just one question to @bilelmoussaoui and @sdroege: should we rename the function? Not sure if `find_with_equal_func` follows the naming conventions...

File: gio/src/list_store.rs [hunk 10]
```diff
@@ -52,6 +60,37 @@ impl<O: IsA<ListStore>> ListStoreExtManual for O {
...
+        item: &P,
+    ) -> Option<u32> {
+        unsafe extern "C" fn equal_func_trampoline(
+            a: glib::ffi::gconstpointer,
+            b: glib::ffi::gconstpointer,
+        ) -> glib::ffi::gboolean {
+            let a: Borrowed<glib::Object> = from_glib_borrow(a as *mut glib::gobject_ffi::GObject);
+            let b: Borrowed<glib::Object> = from_glib_borrow(b as *mut glib::gobject_ffi::GObject);
+        
+            (*a == *b).into_glib()
```
Comment (bilelmoussaoui) [inline gio/src/list_store.rs:77]: Just noticed that this doesn't actually make use of a compare_func. The `find_with_equal_func` is supposed to take a function used to find the item you're looking for. There's already a `find` function that works without having to pass a compare callback. 

The issue we have here and the reason why this wasn't automatically generated I guess is the original C function doesn't provide a way for passing the compare callback as user_data.

Comment (bilelmoussaoui) [reply gio/src/list_store.rs:77]: See https://gitlab.gnome.org/GNOME/glib/-/issues/2447

Comment (sdroege) [reply gio/src/list_store.rs]: You could in theory work around this by storing the closure in TLS but...

File: gio/Gir.toml [hunk 11]
```diff
@@ -704,6 +704,10 @@ manual_traits = ["ListStoreExtManual"]
     name = "sort"
     manual = true
     doc_trait_name = "ListStoreExtManual"
+    [[object.function]]
+    name = "find_with_equal_func"
+    manual = true
+    doc_trait_name = "ListStoreExtManual"
```
Comment (bilelmoussaoui) [inline gio/Gir.toml:710]:
```suggestion
    ignore = true # See https://gitlab.gnome.org/GNOME/glib/-/issues/2447
```

**Review by sdroege — COMMENTED** [review]
Makes sense to me but please squash all commits into one :) There's no point in having 5 commits for this change

**Review by sdroege — APPROVED** [review]

**Pull request merged**